tests/test_query_read_pref_sharded.py

```diff
@@ -55,17 +56,54 @@ def test_query_and_read_mode_sharded(self):
                 with going(next, cursor):
                     request = server.receives()
                     if mode in modes_without_query:
-                        # Query is not edited: {'a': 1} is not nested in $query,
-                        # {'$query': {'a': 1}} is not hoisted.
-                        request.assert_matches(Command(
-                            SON([('find', 'test'),
-                                 ('filter', {'a': 1})])))
+                        # Filter is hoisted out of $query.
+                        request.assert_matches(Command(find_command))
                         self.assertFalse('$readPreference' in request)
                     else:
                         # Command is nested in $query.
                         request.assert_matches(Command(
-                            {'$query': SON([('find', 'test'),
-                                            ('filter', {'a': 1})])}))
+                            SON([('$query', find_command),
+                                 ('$readPreference', mode.document)])))
+
+                    request.replies({'cursor': {'id': 0, 'firstBatch': [{}]}})
+
+    def test_query_and_read_mode_sharded_op_msg(self):
+        """Test OP_MSG sends non-primary $readPreference and never $query."""
+        server = MockupDB()
+        server.autoresponds('ismaster', ismaster=True, msg='isdbgrid',
+                            minWireVersion=2, maxWireVersion=6)
```

Comment:
What if it's *not* a mongos - what if it doesn't say "isdbgrid"? I think OP_MSG requests should include $readPreference to all server types, not just mongos. Let's test that.

Reply:
Added `test_op_msg_read_preference.py` for this. I think the correct thing is to always send $readPreference against all topologies.